datalad_next/tree.py

```diff
@@ -0,0 +1,972 @@
+            dataset.""",
+            constraints=EnsureInt() & EnsureRange(min=0) | EnsureNone()),
+        dataset_depth=Parameter(
+            args=("--dataset-depth",),
+            doc="""maximum level of nested subdatasets to display""",
+            constraints=EnsureInt() & EnsureRange(min=0) | EnsureNone()),
+        include_files=Parameter(
+            args=("--include-files",),
+            doc="""include files in output display""",
+            action='store_true'),
+        include_hidden=Parameter(
+            args=("--include-hidden",),
+            doc="""include hidden files/directories in output display""",
+            action='store_true'),
+    )
+
+    _examples_ = [
+        dict(text="Display up to 3 levels of the current directory's "
+                  "subdirectories and their contents",
+             code_py="tree(depth=3, include_files=True)",
+             code_cmd="datalad tree --depth 3 --include-files"),
+        dict(text="Display all first- and second-level subdatasets of "
+                  "datasets located anywhere under /tmp (including in hidden "
+                  "directories) regardless of directory depth",
+             code_py="tree('/tmp', dataset_depth=2, include_hidden=True)",
+             code_cmd="datalad tree /tmp --dataset-depth 2 --include-hidden"),
+        dict(text="Display first- and second-level subdatasets and their "
+                  "contents, up to 1 directory deep within each dataset",
+             code_py="tree(dataset_depth=2, depth=1)",
+             code_cmd="datalad tree --dataset-depth 2 --depth 1"),
+    ]
+
+    @staticmethod
+    @datasetmethod(name='tree')
+    @eval_results
+    def __call__(
+            path='.',
+            *,
+            depth=None,
+            dataset_depth=None,
+            include_files=False,
+            include_hidden=False):
+
+        if dataset_depth is not None:
+            # special tree defined by subdataset nesting depth
+            tree_cls = DatasetTree
+            dataset_tree_args = {"max_dataset_depth": dataset_depth}
+        else:
+            # simple tree defined by directory depth
+            tree_cls = Tree
+            dataset_tree_args = {}
+
+        tree = tree_cls(
+            Path(path),
+            max_depth=depth,
+            exclude_node_func=build_excluded_node_func(
+                include_hidden=include_hidden, include_files=include_files),
+            **dataset_tree_args
+        )
+
+        for node, line in tree.generate_nodes_with_str():
+            # yield one node at a time to improve UX / perceived speed
+            yield get_status_dict(
+                action="tree",
+                status="ok",
+                path=node.path,
+                type=node.TYPE,
+                depth=node.depth,
+                node_str=line,
+                tree_stats=tree.stats()
+            )
+
+    @staticmethod
+    def custom_result_renderer(res, **kwargs):
+        print(res["node_str"])
```

Comment:
Here and elsewhere, it would be better to use

```py
from datalad.ui import ui
ui.message(...)
```

This takes care of not interfering with other datalad-messaging (e.g. progress bars), and it is also to some degree platform-aware -- it might already help with the windows CI issues that are still present right now.

Reply:
Done. And indeed, it fixed the encoding error on the Windows tests. Awesome!